pyteal/ast/itxn_test.py

```diff
@@ -134,4 +135,126 @@ def test_InnerTxnBuilder_SetFields():
+        {TxnField.fee: pt.Int(0)},
+        pt.Seq(
+            pt.InnerTxnBuilder.SetFields(
+                {
+                    pt.TxnField.type_enum: TxnType.ApplicationCall,
+                    pt.TxnField.application_id: pt.Int(1),
+                    pt.TxnField.accounts: [pt.abi.Account().address()],
+                    pt.TxnField.applications: [pt.abi.Application().application_id()],
+                    pt.TxnField.assets: [pt.abi.Asset().asset_id()],
+                    pt.TxnField.application_args: [
+                        pt.MethodSignature("add(application,asset,account)void"),
+                        pt.Bytes(b"\x01"),
+                        pt.Bytes(b"\x00"),
+                        pt.Bytes(b"\x01"),
+                    ],
+                    pt.TxnField.fee: pt.Int(0),
+                }
+            ),
+        ),
+    ),
+    (
+        pt.Int(1),
+        "add(pay,txn,appl)void",
+        [
+            {TxnField.type_enum: TxnType.Payment},
+            {TxnField.type_enum: TxnType.AssetTransfer},
+            {TxnField.type_enum: TxnType.ApplicationCall},
+        ],
+        {TxnField.fee: pt.Int(0)},
+        pt.Seq(
+            pt.InnerTxnBuilder.SetFields({TxnField.type_enum: TxnType.Payment}),
+            pt.InnerTxnBuilder.Next(),
+            pt.InnerTxnBuilder.SetFields({TxnField.type_enum: TxnType.AssetTransfer}),
+            pt.InnerTxnBuilder.Next(),
+            pt.InnerTxnBuilder.SetFields({TxnField.type_enum: TxnType.ApplicationCall}),
+            pt.InnerTxnBuilder.Next(),
+            pt.InnerTxnBuilder.SetFields(
+                {
+                    pt.TxnField.type_enum: TxnType.ApplicationCall,
+                    pt.TxnField.application_id: pt.Int(1),
+                    pt.TxnField.application_args: [
+                        pt.MethodSignature("add(pay,txn,appl)void"),
+                    ],
+                    pt.TxnField.fee: pt.Int(0),
+                }
+            ),
+        ),
+    ),
+)
+
+
+@pytest.mark.parametrize("app_id, sig, args, fields, expectedExpr", ITXN_METHOD_CASES)
+def test_InnerTxnBuilder_method_call(
+    app_id: pt.Expr,
+    sig: str,
+    args: list[pt.abi.BaseType | pt.Expr | dict[pt.TxnField, pt.Expr | list[pt.Expr]]],
+    fields: dict[pt.TxnField, pt.Expr | list[pt.Expr]],
+    expectedExpr: pt.Expr,
+):
+    expr: pt.Expr = pt.InnerTxnBuilder.MethodCall(
+        app_id=app_id, method_signature=sig, args=args, fields=fields
+    )
+    assert expr.type_of() == pt.TealType.none
+    assert not expr.has_return()
+
+    expected, _ = expectedExpr.__teal__(teal6Options)
+    expected.addIncoming()
+    expected = pt.TealBlock.NormalizeBlocks(expected)
+
+    actual, _ = expr.__teal__(teal6Options)
+    actual.addIncoming()
+    actual = pt.TealBlock.NormalizeBlocks(actual)
+
+    with pt.TealComponent.Context.ignoreScratchSlotEquality(), pt.TealComponent.Context.ignoreExprEquality():
+        assert actual == expected
```

Comment:
When using `TealComponent.Context.ignoreScratchSlotEquality()`, you'll also want to use `TealBlock.MatchScratchSlotReferences` to ensure the scratch slot references are still compatible.

An example's in this docstring: https://github.com/algorand/pyteal/blob/7304e8f07ffc2572a3fff563a625deaf0231f26e/pyteal/ir/tealcomponent.py#L70-L84